self-healing/heal-ldes-data.ts

```diff
@@ -0,0 +1,157 @@
+import { querySudo } from "@lblod/mu-auth-sudo";
+import { sparqlEscapeUri } from "mu";
+import dispatch from "../config/dispatch";
+import {
+  HEALING_DUMP_GRAPH,
+  HEALING_TRANSFORMED_GRAPH,
+  DIRECT_DB_ENDPOINT,
+  HEALING_LIMIT,
+} from "../config";
+import { HealingConfig } from "../config/healing";
+
+export async function healEntities(
+  stream: string,
+  config: HealingConfig
+): Promise<void> {
+  const rdfTypes = Object.keys(config[stream].entities);
+  for (const type of rdfTypes) {
+    const differences = await getDifferences(type, stream, config);
+    await triggerRecreate(differences);
+  }
+}
+
+async function triggerRecreate(differences) {
+  const uniqueSubjects = [
+    ...new Set<string>(differences.map((difference) => difference.s.value)),
+  ];
+  if (uniqueSubjects.length === 0) {
+    console.log("No differences found.");
+    return;
+  }
+  const subjectTypes = await getSubjectTypes(uniqueSubjects);
+  const inserts = subjectTypes.map((s) => {
+    // fake everything but the subject
+    return {
+      subject: { value: s.subject, type: "uri" },
+      predicate: {
+        value: "http://www.w3.org/1999/02/22-rdf-syntax-ns#type",
+        type: "uri",
+      },
+      object: { value: s.type, type: "uri" },
+      graph: { value: "http://mu.semte.ch/graphs/application", type: "uri" },
+    };
+  });
+
+  console.log(
+    `Inserting ${inserts.length} triples: ${JSON.stringify(inserts)}`
+  );
+
+  await dispatch([
+    {
+      inserts,
+      deletes: [],
+    },
+  ]);
+}
+
+async function getSubjectTypes(subjects: string[]) {
+  const result = await querySudo(
+    `
+    SELECT DISTINCT ?s ?type
+    WHERE {
+      VALUES ?s { ${subjects.map(sparqlEscapeUri).join(" ")} }
+      ?s a ?type.
+    }
+  `
+  );
+
+  return result.results.bindings.map((binding) => {
+    return {
+      subject: binding.s.value,
+      type: binding.type.value,
+    };
+  });
+}
+
+async function getDifferences(
+  type: string,
+  stream: string,
+  config: HealingConfig
+) {
+  const predicates =
+    config[stream].entities[type].healingPredicates ||
+    config[stream].entities[type];
```

Comment:
the default is the simple case where it is an array, it can be more complex by adding a healingPredicates list and filter (which is why the second case is there)

Reply:
Does not feel necessary when they have this example 👌🏻